File: vol2/ch15/LeetCode845-2.cpp
class Solution {
public:
    int longestMountain(vector<int>& arr) {
        int n=arr.size();
        if (n<3) return 0;
        vector<int> left(n);
        for (int i=1;i<n;i++) {
            if (arr[i-1]<arr[i])
                left[i]=left[i-1]+1;
            else
                left[i]=0;
        }
        vector<int> right(n);
        for (int j=n-2;j>=0;j--) {
            if(arr[j+1]<arr[j])
                right[j]=right[j+1]+1;
            else
                right[j]=0;
        }
        int ans=0;
        for (int i=0;i<n;i++) {
            if (left[i]>0 && right[i]>0)
                ans=max(ans,left[i]+right[i]+1);
        }
        return ans;
    }
};

File: vol2/ch15/LeetCode845-2.py
class Solution:
    def longestMountain(self, arr: List[int]) -> int:
        n=len(arr)
        if n<3:return 0
        left=[0]*n
        for i in range(1,n):
            left[i]=(left[i-1]+1 if arr[i-1]<arr[i] else 0)       
        right=[0]*n
        for i in range(n-2,-1,-1):
            right[i]=(right[i+1]+1 if arr[i+1]<arr[i] else 0)
        ans=0
        for i in range(n):
            if left[i]>0 and right[i]>0:
                ans=max(ans,left[i]+right[i]+1)
        return ans
